scanpy/tools/_ingest.py

```diff
@@ -0,0 +1,136 @@
+        self._random_init, self._tree_init = make_initialisations(dist_func, dist_args)
+        self._search = make_initialized_nnd_search(dist_func, dist_args)
+
+        search_graph = adata.uns['neighbors']['distances'].copy()
+        search_graph.data = (search_graph.data > 0).astype(np.int8)
+        self._search_graph = search_graph.maximum(search_graph.transpose())
+
+        if 'rp_forest' in adata.uns['neighbors']:
+            self._rp_forest = _rp_forest_generate(adata.uns['neighbors']['rp_forest'])
+        else:
+            self._rp_forest = None
+
+    def __init__(self, adata):
+        #assume rep is X if all initializations fail to identify it
+        self._rep = adata.X
+        self._use_rep = 'X'
+
+        self._n_pcs = None
+
+        #maybe don't need it, rather initialize Ingest class with all needed cluster keys
+        self._adata = adata
+
+        if 'PCs' in adata.varm:
+            self._pca_basis = adata.varm['PCs']
+
+        if 'neighbors' in adata.uns:
+            self._init_neighbors(adata)
+
+        if 'X_umap' in adata.obsm:
+            self._init_umap(adata)
+
+    def pca(self, adata_small, n_pcs=None, inplace=True):
+        #todo - efficient implementation for sparse matrices
+        rep = adata_small.X
+        rep = rep.toarray() if issparse(rep) else rep.copy()
+        rep -= rep.mean(axis=0)
+        X_pca = np.dot(rep, self._pca_basis[:, :n_pcs])
+        if inplace:
+            adata_small.obsm['X_pca'] = X_pca
+        else:
+            return X_pca
+
+    def same_rep(self, adata_small):
+        if self._n_pcs is not None:
+            return self.pca(adata_small, self._n_pcs, inplace=False)
+        if self._use_rep == 'X':
+            return adata_small.X
+        if self._use_rep in adata.obsm.keys():
+            return adata.obsm[self._use_rep]
+        return adata_small.X
+
+    def neighbors(self, adata_small, k=10, queue_size=5, random_state=0):
+        random_state = check_random_state(random_state)
+        rng_state = random_state.randint(INT32_MIN, INT32_MAX, 3).astype(np.int64)
+
+        train = self._rep
+        test = self.same_rep(adata_small)
+
+        init = initialise_search(self._rp_forest, train, test, int(k * queue_size),
+                                 self._random_init, self._tree_init, rng_state)
+
+        result = self._search(train, self._search_graph.indptr, self._search_graph.indices, init, test)
+        indices, dists = deheap_sort(result)
+        return indices[:, :k], dists[:, :k]
+
+    def umap(self, adata_small):
+        rep = self.same_rep(adata_small)
+        adata_small.obsm['X_umap'] = self._umap.transform(rep)
+
+    def knn_classify(self, adata_small, classes_key, k=10, **kwargs):
+        #i.e. ingest.knn_classify(adata_small, 'louvain')
+        cat_array = self._adata.obs[classes_key]
+        neighbors, _ = self.neighbors(adata_small, k, **kwargs)
+
+        values = [cat_array[inds].mode()[0] for inds in neighbors]
```

Comment:
Thinking about this again, what's the reason to include KNN classification in here? Given the large low weight component I've seen in UMAP connectivity graphs, I'm worried this will give noisy results.

Additionally [the recent benchmarking paper](https://www.biorxiv.org/content/10.1101/644435v1) found poor relative performance for the KNN classifier they tried.